### PR TITLE
Add minetest.is_creative_enabled

### LINKED ISSUE #8413: Per-player Creative Mode
##### Issue type
 - Feature request
 
##### Minetest version
```
5.0.0
```

##### The problem
The global setting `creative_mode` is very annoying to use.

For instance, if the server operator wants to fix some block breakage, but the server is not in Creative Mode, they have to shut the server down, turn on Creative Mode, remove a few blocks, start the server again. This sucks.

There are many other situations where the global Creative Mode is just annoying.

##### The solution
Add support for a per-player Creative Mode in the engine!

##### Design
Add a function `minetest.is_in_creative_mode(player)` that returns `true` iff the player is in Creative Mode.

The default implementation will be a simple wrapper around the `creative_mode` setting. Iff `creative_mode=true`, then the function returns true.

Now comes the trick!
Mods are allowed / adviced to overwrite this function to implement a per-player Creative Mode, in a similar fashion like `minetest.is_protected`. Usually this should be done by the game.

The setting `creative_mode` shall continue to exist. However, mods shall be adviced to start using the new function instead of reading from the setting directly. But the change is backwards-compatible: Mods that still continue to read the setting direclty won't break.

If this suggestion is adopted, the current MTG API in `creative` needs to be replaced, obviously.

##### Justification
Yes, I know, this is already implemented in MTG with the function `creative.is_enabled_for`. And yes, every game could already do such an implementation. These implementations would even work.

But this is not good enough for me. This strategy is just not flexible enough and is still annoying for coders.
For instance, if you want to write a game-independent mod that also supports Creative Mode somehow, you can't. You would have to hard-depend on the varying game creative APIs, and not all games even offer these. Dependency hell ensues. You also would have to memorize various different idioms to basically do the same thing. Ugly.

Besides, I always thought of it being kind of ugly for always checking manually the `creative_mode` setting in so many places of my code.

Hence, my suggestion for a **standard** function that all mods can call without worries. Now all a mod needs to do to support Creative Mode is to call a core function that always works without dependency hell.

I think a per-player Creative Mode is so extremely important that IMO it makes much more sense to move it to the engine, because it makes dealing with Creative Mode much less painful, less error-prone and more intuitive.

Some people might argue that a Creative Mode might be too game-specific and should therefore not be supported by the engine at all.
To that I reply that from experience, most games DO have a meaningful Creative Mode and most mods DO want to have some form of Creative Mode support.
It is true there are games where the distinction between Creative Mode on/off does not exist, but those are the exception and my proposal would not hurt them at all. In those exceptional cases, mods would still call `minetest.is_in_creative_mode` and just get the default behaviour returned. So the end result is that the 95% of games that do want a Creative Mode now have a clean standard implementation and for the remaining 5% everything stays the same.

### PR DESCRIPTION
**Pull request by Wuzzy2**
Closes #8413.

This adds a function `minetest.creative_is_enabled_for(name)`.
It returns whether Creative Mode is enabled for the player `name` (can be `""` for non-players).
By default, it returns the value of the `creative_mode` setting (true or false).

This function is designed to be overridden by mods, in a similar fashion like `minetest.is_protected`. It allows games to add a per-player inventory very easily.

This allows games to introduce whatever type of Creative Mode (per-player or not) they want with relative ease.

## To-do
This is a draft for you to look at. Do NOT merge before #9450 is merged, because I plan to make an update after this.

Apart from that, I think it works fine.

## Justification
In one word: Decoupling.

This makes implementing a per-player Creative Mode a piece of cake. Without this, mods always had to rely on some game API, which of course will differ between games.

Using `minetest.creative_is_enabled_for` will become the preferred way to check for Creative Mode.

## Game updates
This section is informational only, it talks about what to do with games afterwards.

### Minetest Game update
Minetest Game already implements a per-player Creative Mode, but mods need to hard-depend on a mod to use it. This implementation will obviously need updating after this PR. If this MTG update has been done, mods can start to use `minetest.creative_is_enabled_for` without needing to depend on anything.

### Updates for other games and mods
Basically, all they have to do is to replace all instances of `minetest.settings:get_bool("creative_mode")` with `minetest.creative_is_enabled_for`.

### CODE REVIEW

File: builtin/common/misc_helpers.lua [hunk 1]
```diff
@@ -344,17 +344,11 @@ if INIT == "game" then
 --Wrapper for rotate_and_place() to check for sneak and assume Creative mode
 --implies infinite stacks when performing a 6d rotation.
 --------------------------------------------------------------------------------
-	local creative_mode_cache = core.settings:get_bool("creative_mode")
-	local function is_creative(name)
-		return creative_mode_cache or
-				core.check_player_privs(name, {creative = true})
-	end
```
Comment (MoNTE48) [inline builtin/common/misc_helpers.lua:351]: Any reason to remove privilege check and force me to override this function to return creative for only 1 player (me)  on my survival server?

**Comment by MoNTE48** [comment]
Hey, at least one reason why this has not been done before?
It should be merged as soon as possible!  👍

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5440,6 +5440,13 @@ Misc.
 * `minetest.record_protection_violation(pos, name)`
     * This function calls functions registered with
       `minetest.register_on_protection_violation`.
+* `minetest.creative_is_enabled_for(name)`: returns boolean
+    * Returning `true` means that Creative Mode is enabled for player `name`.
+    * `name` will be `""` for non-players or unknown players.
```
Comment (sfan5) [inline doc/lua_api.txt:5445]:
```suggestion
    * `name` will be `""` for non-players or if the player is unknown.
```

File: builtin/common/misc_helpers.lua (hunk 1 above)
Comment (sfan5) [reply builtin/common/misc_helpers.lua:351]: Is the `creative` privilege documented anywhere? If not, the engine should not have code to support it.

File: builtin/game/misc.lua [hunk 3]
```diff
@@ -164,6 +164,12 @@ function core.record_protection_violation(pos, name)
 	end
 end
 
+-- To be overridden by Creative mods
+
+local creative_mode_cache = core.settings:get_bool("creative_mode", false)
+function core.creative_is_enabled_for(name)
```
Comment (sfan5) [inline builtin/game/misc.lua:170]: I propose renaming this to `creative_is_enabled` (or maybe `creative_enabled`). The naming with a verb and `_for()` is not used anywhere else.

**Comment by Wuzzy2** [comment]
`creative` is not even a builtin privilege, so IMO it doesn't make sense to check it in builtin. Also, builtin checked the `creative` priv inconsistently (only in this code fragment, in other places, only the `creative_mode` setting was checkd).

That's why I decided to replace this fragment.

File: builtin/common/misc_helpers.lua (hunk 1 above)
Comment (MoNTE48) [reply builtin/common/misc_helpers.lua:351]: Then let's document it. This is really a trivial issue.

Comment (rubenwardy) [reply builtin/common/misc_helpers.lua:351]: The engine doesn't have a creative priv. That's in the creative mod. The creative mod can just override this to add the creative priv

Comment (MoNTE48) [reply builtin/common/misc_helpers.lua:351]: @rubenwardy, so, the engine supports per-player immortal (starting 5.1), per-player sky, per-player time, per-player FOV, etc.  As well as two dozen vaunting privileges.
But per-player creative should strictly add each subgame by each developer?

**Comment by Panquesito7** [comment]
The title is outdated, it shall be `minetest.creative_is_enabled` now. 🙂

File: builtin/common/misc_helpers.lua [hunk 4]
```diff
@@ -344,17 +344,11 @@ if INIT == "game" then
...
-
 	core.rotate_node = function(itemstack, placer, pointed_thing)
 		local name = placer and placer:get_player_name() or ""
 		local invert_wall = placer and placer:get_player_control().sneak or false
 		return core.rotate_and_place(itemstack, placer, pointed_thing,
-				is_creative(name),
+				core.creative_is_enabled(name),
```
Comment (Panquesito7) [inline builtin/common/misc_helpers.lua:351]: Too much indentation.
```suggestion
			core.creative_is_enabled(name),
```

**Comment by rubenwardy** [comment]
I'm not convinced by this PR, and I think it's going in the wrong direction. The engine should not know about creative at all, the only reason it does currently is to show a checkbox and apparently affect rotations.

The creative mod is the decoupled API for checking whether a player is in creative. A game like mineclone can provide a creative mod with a different implementation but implementing the same API.

I don't like the creative privilege, it violates the principle of having privileges define capability rather than state.

I'm not opposed enough to reject this PR, but did want to state my concerns.

> per-player immortal (starting 5.1), per-player sky, per-player time, per-player FOV, etc.

None of those features affect gameplay or force anything on games that don't need them

**Comment by rubenwardy** [comment]
> The creative mod is the decoupled API for checking whether a player is in creative

The creative mod does have additional methods, which makes it inappropriate for a shared API. APIs between games are a complete mess at the moment, it would be nice to have method that didn't involve adding engine support for everything

---

Add WIP label due to OP comment about this being a draft

**Comment by Wuzzy2** [comment]
About the draft status: Note this PR is basically finished, I just wait for #9450 because I can then update this PR to update the handling of Creative Mode in Minimal. Apart from that, I do not plan other updates (for now).

So although this PR is *technically* a draft, it is very ready for review.

---

Personally, I think we have Creative Mode mostly out of convention/convenience. A LOT of games do have Creative Mode, and have depended on the `creative_mode` setting, which is an OFFICIAL setting. If I take your argumentation at face value, we would have to remove the `creative_mode` setting completely. Which would be an extreme step which would annoy a LOT of players. Yes, not ALL games need Creative Mode, but when did this kind of argument ever stop an feature from being added. Just because something is not used by all games does not mean it's worthless. It's still used my, let's say, most games.
It's good to have *some* kind of standardization.

I see my PR really just like an interface like in Java.

Just like we have `minetest.is_protected`. Minetest is agnostic about how games implement a protection mode or even if they have any at all. In some games, protection might be completely pointless. But if they DO have protection, this function template is a massive help. So if we really take your logic to its conclusion, then `minetest.is_protected` has to be removed as well. No, just no.

Not ALL games need protection, but `minetest.is_protected` is invaluable for those games that need it. Protection is a feature common enough to have made it into the engine. It's the exact same with Creative Mode.

> The creative mod is the decoupled API for checking whether a player is in creative.

Your MTG bias is showing here. No, it's not decoupled. It's coupled into Minetest Game. That's like adding the protection API into Minetest Game and then using this as an argument to prevent `minetest.is_protected` from being added because it's already in MTG …
And you even refute yourself a few sentences later by saying that it is “inappropriate for a shared API” … You're not making a lot of sense here. :-(

> APIs between games are a complete mess at the moment, it would be nice to have method that didn't involve adding engine support for everything

Just because I added this API for ONE specific feature does not mean that we have to do this for every conceivable feature. Slippery slope fallacy, maybe?

Note that Creative Mode is used by a TON of games for Minetest, so it's not just any random feature I'm trying to push here. Your reasoning might be correct for more obscure features, but this isn't the case here. Note that the actual implementation of Creative Mode is still up to the game. If a game does not have a Creative Mode, this PR is completely harmless and does not do any damage.

I believe you are just trying to make the engine as “pure” as possible. I understand the idea, and I generally agree, but when you take it too far, you end up with an engine that is not very useful, forcing games to redundantly re-implement core functionality over and over again. And in the case of Creative Mode, this is the kind of evil I want to overcome. Enabling games to get rid of redundant implementations is a *good* thing.

I can't think of a better solution of doing a per-player Creative Mode than like this. It saves games annoying extra code and also avoids silly hacks like the `creative` privilege in MTG. Note this privilege is MTG's fault, not mine.

(Besides, we're actually talking about `builtin` here, not the engine, but you know what I mean.)

> The creative mod does have additional methods, which makes it inappropriate for a shared API.

That's why a shared, standardized API is so much more desirable.

**Comment by Wuzzy2** [comment]
Seriously, we are arguing all this about a SINGLE function that is completely optional, does not harm any existing game when they don't need it and will help in cleaning up Creative Mode implementations across many games (if they want that).

>  A game like mineclone can provide a creative mod with a different implementation but implementing the same API.

Yeah, no. It's going to be VERY hard to convince game authors to use the same mod just for the API … There will be competing implementation, and it's a mess. It will be the same mess we had with protection mods before we had `minetest.is_protected`. `minetest.is_protected` was a serious improvement.

I'm not saying we should go the “interface” way for every conceivable problem, as it very well will blow up our code. But in the case of Creative Mode I think it really is the best way.

What I only agree about Creative Mode is that Minetest should not provide any special hardcoded gameplay features related to Creative Mode. Currently, that's almost true, the only actual feature that Minetest implements is disabling wear when Creative Mode is enabled. That's wrong, and should be left to the games. If this is removed, Creative Mode becomes “pure” in Minetest. But I digress …

Also: As if this is only about MCL2!
This will be useful to countless games that want to have a Creative Mode.

File: builtin/game/misc.lua [hunk 5]
```diff
@@ -164,6 +164,12 @@ function core.record_protection_violation(pos, name)
 	end
 end
 
+-- To be overridden by Creative mods
+
+local creative_mode_cache = core.settings:get_bool("creative_mode", false)
```
Comment (sfan5) [inline builtin/game/misc.lua:169]: the `, false` is not needed here btw
the engine has a default value for it

**Comment by rubenwardy** [comment]
Thanks for the arguments, I generally support this now

**Comment by Wuzzy2** [comment]
Is this an official approval? Please add a label then. Thx.

**Comment by sorcerykid** [comment]
Minor picky detail, but can this please be renamed to `minetest.is_creative_enabled()`? That would be more consistent with how logic-related functions are typically named. The other name insinuates that creative is always enabled :)

**Comment by Wuzzy2** [comment]
UPDATE!

* Function renamed to `minetest.is_creative_enabled` (Warning: issue title and description are NOT updated yet!)
* PR adjusted for Development Test [devtest]
* PR ready for merge

**Comment by paramat** [comment]
After seeing SmallJoker considering this on IRC, i want to suggest considering a more descriptive API name while we have the chance.
The current name does not mention players and slightly suggests something different.
I do not think we can rely on the documented argument `name` to make it clear.
I do not feel strongly about this though.
`minetest.is_creative_enabled_for(name)` ?
`minetest.is_player_creative(name)` ?
Or something else?
@sfan5 @rubenwardy @SmallJoker

**Comment by rubenwardy** [comment]
> `minetest.is_player_creative(name)`

How rude, all of our players are creative

**Comment by sfan5** [comment]
I don't think a name change is necessary but I'm not strongly opposed either if other devs feel it's necessary.

**Comment by SmallJoker** [comment]
@paramat If this function would be per-player, the function name would need to contain a "player" part. However, it also returns the global creative setting value. I think it's okay as-is.

**Comment by paramat** [comment]
> However, it also returns the global creative setting value.

I did not realise that.
Ok, i no longer request.
